auth_token/models.py

```diff
@@ -129,55 +171,357 @@ def delete(self):
+        return check_password(str(token), self.login_token)
+
+
+class VerificationTokenManager(models.Manager):
+
+    def deactivate(self, obj, slug=None, key=None):
+        self.filter_active_tokens(obj, slug, key).update(is_active=False)
+
+    def deactivate_and_create(self, obj, slug=None, extra_data=None, deactivate_old_tokens=True,
+                              key_generator_kwargs=None, **kwargs):
+        if deactivate_old_tokens:
+            self.deactivate(obj, slug)
+
+        return self._create(obj, slug=slug, extra_data=extra_data, key_generator_kwargs=key_generator_kwargs, **kwargs)
+
+    def get_active_or_create(self, obj, slug=None, extra_data=None, key=None, key_generator_kwargs=None, **kwargs):
+        token = self.filter_active_tokens(obj, slug, key).order_by('created_at').last()
+
+        if token and token.is_valid:
+            return token
+        else:
+            return self._create(obj, slug=slug, extra_data=extra_data, key_generator_kwargs=key_generator_kwargs,
+                                **kwargs
+            )
+
+    def _create(self, obj, slug=None, extra_data=None, key_generator_kwargs=None, **kwargs):
+        expiration_in_minutes = kwargs.pop('expiration_in_minutes', settings.DEFAULT_EXPIRATION)
+        key_generator_kwargs = {} if key_generator_kwargs is None else key_generator_kwargs
+
+        token = self.model(
+            content_type=ContentType.objects.get_for_model(obj.__class__),
+            object_id=obj.pk,
+            slug=slug,
+            expires_at=(timezone.now() + timedelta(minutes=expiration_in_minutes)) if expiration_in_minutes else None,
+            key=self.model.generate_key(**key_generator_kwargs),
+        )
+        if extra_data:
+            token.set_extra_data(extra_data)
+
+        token.save()
+        return token
+
+    def exists_valid(self, obj, key, slug=None):
+        for token in self.filter_active_tokens(obj, slug):
+            if token.check_key(key):
+                return True
+        return False
+
+    def filter_active_tokens(self, obj_or_class, slug=None, key=None):
+        qs = self.filter(
+            is_active=True,
+            slug=slug,
+        )
+
+        if isinstance(obj_or_class, models.Model):
+            qs = qs.filter(object_id=obj_or_class.pk)
+        return qs.filter(key=key) if key else qs
+
+
+class OneTimePasswordManager(SmartManager):
+
+    def create(self, key_generator, **kwargs):
+        for attempt in range(settings.MAX_RANDOM_KEY_ITERATIONS + 1):
+            try:
+                key = key_generator()
+                hashed_key = hash_key(key)
+                obj = super().create(key=hashed_key, **kwargs)
+                obj.secret_key = key
+                return obj
+            except IntegrityError:
+                if attempt > settings.MAX_RANDOM_KEY_ITERATIONS:
+                    raise IntegrityError('Could not produce unique key for authorization token')
+
+
+class OneTimePassword(SmartModel):
```

Comment:
I think there is a naming discrepancy - either use OneTimePassword or VerificationToken.